### PR TITLE
feat: add job refresh

### PR DESCRIPTION
**Pull request by arinda-arif**
https://github.com/odpf/optimus/issues/180

Acceptance criteria:
- Should able to refresh:
    - selected jobs, or 
    - all jobs in selected namespaces, or 
    - all jobs in a project
- After doing refresh, ALL jobs in the requested project will be deployed
- Dependency should be correctly persisted
- If dependency resolution is failed for a job, it should not failed the whole refresh process
- Users should know which jobs failed on refresh or deploy
- Metrics on every process in refresh should be pushed

### CODE REVIEW

**Comment by arinda-arif** [comment]
https://github.com/odpf/proton/pull/106

File: store/postgres/migrations/000020_add_job_dependencies_table.down.sql
Comment (sbchaos) [inline store/postgres/migrations/000020_add_job_dependencies_table.down.sql:1]: We should be dropping table `job_dependency`.

File: store/store.go
Comment (sbchaos) [inline store/store.go:46]: Please create a new model in Project for the `ProjectID` and use it in here.

File: store/store.go
Comment (sbchaos) [inline store/store.go:125]: Use model for `ProjectID` here, so that both do not get mixed.

File: store/postgres/migrations/000020_add_job_dependencies_table.up.sql
Comment (sbchaos) [inline store/postgres/migrations/000020_add_job_dependencies_table.up.sql:4]: Should we create a ForeignKey constraint for `project_id`.

File: store/postgres/job_spec_repository.go
Comment (sbchaos) [inline store/postgres/job_spec_repository.go:68]: Are we able to query multiple jobs using `First` ?

As per gorm docs:
GORM provides First, Take, Last methods to retrieve a single object from the database, it adds LIMIT 1 condition when querying the database

File: store/postgres/job_dependency_repository_test.go
Comment (sbchaos) [inline store/postgres/job_dependency_repository_test.go:17]: The name of function should be `TestIntegrationJobDependencyRepository`, we use name TestIntegration to identify the integration tests.

File: store/postgres/job_dependency_repository_test.go
Comment (sbchaos) [inline store/postgres/job_dependency_repository_test.go:46]: This should not be required, now we run the migrations once before any test is run.
Please use following in place of code above
```
		dbConn := setupDB()
		truncateTables(dbConn)
```

File: store/postgres/job_dependency_repository_test.go
Comment (sbchaos) [inline store/postgres/job_dependency_repository_test.go:56]: These 2 lines should not be required after the above change.

File: store/postgres/job_dependency_repository_test.go
Comment (sbchaos) [inline store/postgres/job_dependency_repository_test.go:60]: Please use `uuid.New()` which is same thing.

File: store/postgres/job_spec_repository.go
Comment (sbchaos) [inline store/postgres/job_spec_repository.go:66]: Please add test for this method.

File: store/postgres/job_dependency_repository.go
Comment (sbchaos) [inline store/postgres/job_dependency_repository.go:28]: Please do not use project in repository, pass it as a method parameter.

File: cmd/server/server.go
Comment (sbchaos) [inline cmd/server/server.go:181]: Can we get rid of repoFactory and use the repo directly.

File: store/postgres/migrations/000020_add_job_dependencies_table.down.sql
Comment (arinda-arif) [reply store/postgres/migrations/000020_add_job_dependencies_table.down.sql]: updated

File: store/postgres/migrations/000020_add_job_dependencies_table.up.sql
Comment (arinda-arif) [reply store/postgres/migrations/000020_add_job_dependencies_table.up.sql]: added for both `project_id` and `dependent_project_id`. also have created index for `job_id` and `project_id`.

File: store/postgres/job_spec_repository.go
Comment (arinda-arif) [reply store/postgres/job_spec_repository.go]: you are right, instead of `First` this should use `Find`. updated.

File: store/postgres/job_dependency_repository_test.go
Comment (arinda-arif) [reply store/postgres/job_dependency_repository_test.go]: updated

File: store/postgres/job_dependency_repository_test.go
Comment (arinda-arif) [reply store/postgres/job_dependency_repository_test.go]: updated

File: store/postgres/job_dependency_repository_test.go
Comment (arinda-arif) [reply store/postgres/job_dependency_repository_test.go]: removed

File: store/postgres/job_dependency_repository_test.go
Comment (arinda-arif) [reply store/postgres/job_dependency_repository_test.go]: updated

File: store/postgres/job_spec_repository.go
Comment (arinda-arif) [reply store/postgres/job_spec_repository.go:66]: test added

File: cmd/server/server.go
Comment (arinda-arif) [reply cmd/server/server.go]: removed

File: store/postgres/job_dependency_repository_test.go
Comment (arinda-arif) [reply store/postgres/job_dependency_repository_test.go]: also need to add to truncateTables list

File: store/store.go
Comment (arinda-arif) [reply store/store.go]: deleted `GetByID` from project repository as it is only being used inside `project_repository_test`. to verify the test, i have replaced `GetByID` with `GetByName`.

File: store/store.go
Comment (arinda-arif) [reply store/store.go]: have added models.ProjectID and use it in ProjectSpec

**Comment by arinda-arif** [comment]
Hi @sbchaos, thank you for your feedback.

I have made the changes to resolve them. Please help to retake a look.

File: job/service.go
Comment (sbchaos) [inline job/service.go:86]: Can we please remove this

File: models/project.go
Comment (sbchaos) [inline models/project.go:42]: 👍

**Review by sbchaos — APPROVED** [review]

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: removed

File: job/service.go
Comment (sravankorumilli) [inline job/service.go:710]: Similarly here, as you are refetching the dependencies which were constructed above i believe your intention is to move to seperate place later. It would be better to do that right now.
1. FetchAllJobsAgain
2. FetchAllJobDependencies
3. EnrichJobsWithJobDependencies
4. EnrichJobsWithHookDependencies // This needs to be optimized by not making call to plugin inside enrichment rather we cache the dependencies between plugins during bootstrap and use pluginService for these kind of purposes.
5. PriorityResolution 
6. Grouping Namespecs // This step too, why are we keep on passing the projectJobSpecRep and fetching jobs lets just enrich the JobSpecModel with Project & NamespaceSpec or store ProjectId and NamespaceID in the JobSpec which can be used for grouping.

I expect all the data from datasources or third party integrations to happen only once and this data can be cached for optimization purposes. in here we are passing jobspecrepo in every place and fetching jobspecs again and again which can be avoided.

File: job/service.go
Comment (sravankorumilli) [inline job/service.go:852]: prepare is generic better to have specific functions so readers can easily understand whats happening

File: job/service.go
Comment (sravankorumilli) [inline job/service.go:769]: If possible we can see a mechanism on abstract out all the operations of how the parallel running is happening by keeping things declarative

File: job/dependency_resolver.go
Comment (sravankorumilli) [inline job/dependency_resolver.go:98]: FetchJobDependencies I believe can be simplified the logic of fetching and updating with inter/intra job dependencies .

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: job/service.go
Comment (arinda-arif) [reply job/service.go:53]: agree, I think it will reduce the complexity. however, in the existing resolve function, there is a `resolveHookDependencies` process, which is not needed (as we are going to store the requested and fetch all after). should we also decouple it from `Resolve` or keep it until we refactor job `Sync`?

File: job/service.go
Comment (arinda-arif) [reply job/service.go:698]: I was limiting the usage of protobuf model to only in the handler and not passing it to the service layer. this `namespaceJobNamePairs ` is only being used to get all of the jobSpecs for dependency resolution. should we do it early in the handler? what do you think?

File: job/service.go
Comment (arinda-arif) [reply job/service.go:703]: there are several possible errors that happens in that particular function: multierror (errors from dependency resolution), when compiling assets, or when fetching the job specs. if it is from dependency resolution, we should skip the error, as we don't want the whole refresh process to be canceled. looks like it is not clear enough, i will refactor it.

File: job/service.go
Comment (arinda-arif) [reply job/service.go:701]: agree, will change.

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: have changed it to only a single gauge. differentiate the metric values with `status` label.

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: As discussed, reuse resolve and introduce `Persist` instead of `ResolveAndPersist`

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: As discussed, still use `NamespaceJobNamePair` to avoid using external contract in the service layer. However, we are not passing NamespaceSpec, but NamespaceName instead (and fetching the spec just before where it is being used).

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: Have modified this so it will not check for multierror, it will check for the error type.

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: modifiedthis, however persist dependencies and resolve dependencies are still in a same parallel runner. please check this.

File: api/handler/v1beta1/job_spec.go
Comment (arinda-arif) [reply api/handler/v1beta1/job_spec.go]: as discussed, we will still use namespaceJobNamePairs to avoid using external contract in service layer.

File: models/job.go
Comment (arinda-arif) [reply models/job.go:436]: The values of JobSpecDependencyType are in the same file, in the const section (top of the file). This should be enough?

File: job/dependency_resolver.go
Comment (arinda-arif) [reply job/dependency_resolver.go]: As discussed, we are still using uuid for now, to avoid having very big change in 1 PR

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: updated. please help to recheck

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: renamed this to `fetchJobSpecs`

File: job/dependency_resolver.go
Comment (arinda-arif) [reply job/dependency_resolver.go]: have modified this, FetchJobDependencies will just fetch the dependencies, adapting to the required model is handled on the enriching part (in deployer).

File: job/service.go
Comment (arinda-arif) [reply job/service.go:793]: decided to not refactor this part in this PR, as abstracting this out is quite tricky, and will see a better way to do this.

File: cmd/job_refresh.go
Comment (sravankorumilli) [inline cmd/job_refresh.go:127]: errors can be warn messages

File: cmd/job_refresh.go
Comment (sravankorumilli) [inline cmd/job_refresh.go:141]: failures to be considered as warn logs

File: cmd/job_refresh.go
Comment (sravankorumilli) [inline cmd/job_refresh.go:97]: this nested if can be merged by just using single if on errors.Is

File: api/handler/v1beta1/job_spec.go
Comment (sravankorumilli) [inline api/handler/v1beta1/job_spec.go:302]: can you split this into multiple lines, will be more readable

File: models/namespace.go
Comment (sravankorumilli) [inline models/namespace.go:20]: we don't need this struct, we can avoid this.

File: api/handler/v1beta1/job_spec.go
Comment (sravankorumilli) [reply api/handler/v1beta1/job_spec.go:302]: Can you change the proto request to accept request for project refresh or bunch of namespaces in project or bunch of jobs. project is mandatory. namespaces and jobs list is optional.
We can have different functions for refreshing all jobs for a project. refresh all jobs of given namespaces, refresh bunch of jobs in a project.

File: job/service.go
Comment (sravankorumilli) [inline job/service.go:737]: fetchJobSpecs should be split into 4 functions, fetchAllForAProject, fetchAllForGivenNamespaces, fetchSpecsForGivenJobNames,  fetchJobSpecs will be the wrapper function

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: api/handler/v1beta1/job_spec.go
Comment (arinda-arif) [reply api/handler/v1beta1/job_spec.go:302]: is this what you meant?
```proto
message RefreshJobsRequest {
  string project_name = 1;
  repeated string namespace_names = 2;
  repeated string job_names = 3;
}
```
currently, to refresh a bunch of jobs in a project, we need namespace info of where the jobs belong (for auth). we can go with what you proposed, but will still require the namespace that needs to be provided in the `namespace_names` (which is not straightforward), or we can avoid requesting the namespace for this case.

File: cmd/job_refresh.go
Comment (arinda-arif) [reply cmd/job_refresh.go:97]: the reason behind having nested if in error is to give more information for the specific error (in this case time out), besides showing a message of other errors. what do you think?

File: cmd/job_refresh.go
Comment (arinda-arif) [reply cmd/job_refresh.go:113]: the intention of having nested ifs is to ignore EOF and throw other errors. another option is to not make nested, but multiple ifs that are quite similar.

File: job/deployer.go
Comment (sravankorumilli) [inline job/deployer.go:73]: we can refactor this method
1. createJobSpecMap can be dissolved into fetchJobSpecsForDependentJobs which can go into the main function call and the map creation will just  create JobSpecMap given jobSpecs and dependentJobSpecs, here we can optimize by grouping the jobs by dependent project and fetching only once. Or why not fetch the jobspecs through a foreign key contraint, then we can avoid this fetch altogether.
2. We will have a function to enrich a single obSpec With JobDependencies, which will be reused here. When ever there is nesting better to break down into multiple functions

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: cmd/job_refresh.go
Comment (arinda-arif) [reply cmd/job_refresh.go]: updated

File: cmd/job_refresh.go
Comment (arinda-arif) [reply cmd/job_refresh.go]: updated

File: api/handler/v1beta1/job_spec.go
Comment (arinda-arif) [reply api/handler/v1beta1/job_spec.go]: updated, the proto PR: https://github.com/odpf/proton/pull/127

File: models/namespace.go
Comment (arinda-arif) [reply models/namespace.go]: removed this as we are no longer need namespace to refresh specific jobs.

File: job/service.go
Comment (arinda-arif) [reply job/service.go]: updated

File: job/deployer.go
Comment (arinda-arif) [reply job/deployer.go]: updated as discussed: not fetching the jobspecs through foreign key constraint (to avoid performance issue), jobSpecMap and enrichment are now being done when fetching job specs with job dependencies.

File: job/service.go
Comment (sravankorumilli) [inline job/service.go:726]: why fail in case of asset compilation

File: cmd/job_refresh.go
Comment (sravankorumilli) [inline cmd/job_refresh.go:35]: looks like we are accepting the flags but it doesn't work, as we are always referring from config. @deryrahman is working on fixing loading of specs from a single place, lets do the same here.

**Review by sravankorumilli — CHANGES_REQUESTED** [review]

File: cmd/job_refresh.go
Comment (deryrahman) [reply cmd/job_refresh.go:35]: yes, it's known issue @sravankorumilli , will be fixed on https://github.com/odpf/optimus/pull/274, let's merge it once the PR is approved

File: job/service.go
Comment (arinda-arif) [reply job/service.go:726]: this should be removed by now. fixing

Comment (arinda-arif) [reply job/service.go]: removed

File: cmd/job_refresh.go
Comment (arinda-arif) [reply cmd/job_refresh.go]: for now i have removed the project flag

**Review by sravankorumilli — APPROVED** [review]

**Pull request merged**
